crates/shielded_token/src/masp/shielded_wallet.rs

```diff
@@ -2272,22 +2782,27 @@ mod test_shielded_wallet {
                     vk,
                 );
 
+                let balance = wallet
+                    .compute_shielded_balance(&vk)
+                    .await
+                    .unwrap()
+                    .unwrap_or_else(ValueSum::zero);
                 let rewards_est = wallet
-                    .estimate_next_epoch_rewards(&context, &vk)
+                    .estimate_next_epoch_rewards(&context, &balance)
                     .await
                     .expect("Test failed");
-                let principal1 = i128::from(principal1);
-                let principal2 = i128::from(principal2);
-                // reward from epoch 1->2 + reward from epoch 2->3 + reward from
-                // epoch 2->3
-                let expected_tok1_rewards =
-                    principal1 * tok1_reward_rate + principal1 + principal1;
-                // reward from epoch 2->3 + reward from epoch 2->3
-                let expected_tok2_rewards =
-                    principal2 * tok2_reward_rate + principal2 * tok2_reward_rate;
+                // The native token balances at epoch 3 and 4 are given by the shielded amounts times the rewards times the conversion for the native asset
+                let native_balance_at_3 = (i128::from(principal1) * (tok1_reward_rate + 1) + i128::from(principal2) * tok2_reward_rate) * 2;
```

Comment:
Wait, maybe I've got it. Is this supposed to return the balance, i.e. the principal + rewards? If so, then these mysterious `x+` make sense.

Reply:
Only in this case, for the previous two cases no, it's supposed to return the estimated rewards

Reply:
🙈